python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -482,17 +509,93 @@ def associate(self, collection, refs):
         """Add existing Datasets to a collection, possibly creating the
         collection in the process.
 
+        If a DatasetRef with the same exact `dataset_id`` is alredy in a
+        collection nothing is changed. If a DatasetRef with the same
+        DatasetType and unit values but with different ``dataset_id`` exists
+        in a collection then exception is raised.
+
         Parameters
         ----------
         collection : `str`
             Indicates the collection the Datasets should be associated with.
         refs : `list` of `DatasetRef`
             A `list` of `DatasetRef` instances that already exist in this
             `SqlRegistry`.
+
+        Raises
+        ------
+        ValueError
+            If a Dataset with the given `DatasetRef` already exists in the
+            given collection.
         """
-        datasetCollectionTable = self._schema.tables["DatasetCollection"]
-        self._connection.execute(datasetCollectionTable.insert(),
-                                 [{"dataset_id": ref.id, "collection": collection} for ref in refs])
+        # A collection cannot contain more than one Dataset with the same
+        # DatasetRef. Our SQL schema does not enforce this constaint yet so
+        # checks have to be done in the code:
+        # - read existing collection and try to match its contents with
+        #   new DatasetRefs using units
+        # - if there is a match and dataset_id is different then constraint
+        #   check fails
+        # TODO: This implementation has a race which can violate the
+        # constraint if multiple clients update registry concurrently. Proper
+        # constraint checks have to be implmented in schema.
+
+        def _matchRef(row, ref):
+            """Compare Dataset table row with a DatasetRef.
+
+            Parameters
+            ----------
+            row : `sqlalchemy.RowProxy`
+                Single row from Dataset table.
+            ref : `DatasetRef`
+
+            Returns
+            -------
+            match : `bool`
+                True if Dataset row is identical to ``ref`` (their IDs match),
+                False otherwise.
+
+            Raises
+            ------
+            ValueError
+                If DatasetRef unit values match row data but their IDs differ.
+            """
+            if row.dataset_id == ref.id:
+                return True
+
+            columns = self._dataUnits.getPrimaryKeyNames(ref.datasetType.dataUnits)
+            dataId = ref.dataId
+            if all(row[col] == dataId[col] for col in columns):
```

Comment:
I think the columns compared here also needs to include `dataset_type_name`.

Reply:
I knew something was missing :grin: I'll try to make a unit test for this.